Data/Functor/Foldable/TH.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE TemplateHaskell #-}
+module Data.Functor.Foldable.TH
+  ( makeBaseFunctor
+  ) where
+
+import Data.Bifunctor (first)
+import Data.Functor.Foldable
+import Language.Haskell.TH
+
+makeBaseFunctor :: Name -> DecsQ
```

Comment:
`makeBaseFunctor` seems woefully underdocumented. From the test suite you included, it seems to do four (?) things:

1. Generate a new datatype (with an `F` suffix)
2. Derive a `Base` instance between the supplied `Name` and the new datatype
3. Derive a `Recursive` instance
4. Derive a `Corecursive` instance

Is that correct? Documenting what exactly `makeBaseFunctor` does—and including a small example of what happens when you call `makeBaseFunctor` on a simple datatype in the Haddocks—would be tremendously helpful.

Some other questions:

* Will this fail in the event that there's already an `F`-suffixed name in scope? For example, if you already have `Expr` and `ExprF`, and you also include `$(makeBaseFunctor ''Expr)`? If so, is there a way to override the default behavior to give it a different suffix (perhaps through an `Options` datatype, _à la_ `aeson`)?
* Are there limitations to `makeBaseFunctor` based on the structure of the datatype? Does it handle things like `GADTs`? `RankNTypes`? `ExistentialQuantification`? Polymorphic recursion? Data families? If not, it'd be nice to document when `makeBaseFunctor` will fail.